### DISCUSSION TITLE
Understanding usage of monaco-vscode-api

### DISCUSSION THREAD

**Discussion by dlvhdr** [TypeFox/monaco-languageclient#404] [General]
Hi!

I'm trying to understand how `monaco-languageclient` uses monaco-vscode-api.
Following the code, I'm not sure I understand.
My main gap is around the role of the `extHost` file in `monaco-vscode-api`.
What I'm seeing in the bundled code differs from what's inside [languages.ts](https://github.com/CodinGame/monaco-vscode-api/blob/main/src/vscode-services/languages.ts#L76)

If I put a `debugger;` statement I can see the stack trace as follows:
<img src="https://user-images.githubusercontent.com/6196971/180171422-66625ffe-cfdb-4570-ab2a-85a4ca5a0737.png" width="500" />

The code looks like this:

1. `registerProvider` (in `vscode-languageclient`):

```ts
registerProvider(selector, provider) {
        return vscode_1.languages.registerHoverProvider(this._client.protocol2CodeConverter.asDocumentSelector(selector), provider);
    }
```
2. `registerHoverProvider` (in `monaco-vscode-api`):

```ts
registerHoverProvider(documentSelector, provider) {
        const adapter = new HoverAdapter(extHostDocuments, provider);
        return monaco.languages.registerHoverProvider(documentSelector, {
            provideHover: (model, position, token) => {
                return adapter.provideHover(model.uri, position, token);
            }
        });
    },
```

So, my questions are:
1. What is the role of the `extHost.ts` file?
2. Why am I not seeing it called in the bundled files? My initial intuition was that you'd simply call `monaco.languages` api like in the actual stack trace.
3. I thought I understood that you simply replace the `vscode` import with `monaco-vscode-api` but I'm seeing a lot of code in the `rollup` config etc. How does that work?

Thanks!

**Comment by CGNonofr** [comment]
Hello!

It just seems you use an old version of monaco-vscode-api that was not using the extHost stuff

see https://github.com/TypeFox/monaco-languageclient/pull/399

Btw, you can still probably force a newer version of monaco-vscode-api (be careful, the last versions are the 1.67.x, not the 1.68.x, and we can't unpublish npm packages...)

> What is the role of the extHost.ts file?

It gathers code from vscode, the extHost classes are the "thing" vscode use between its api and the services

> Why am I not seeing it called in the bundled files? My initial intuition was that you'd simply call monaco.languages api like in the actual stack trace.

Because of the wrong version

> I thought I understood that you simply replace the vscode import with monaco-vscode-api

You are right!

>  but I'm seeing a lot of code in the rollup config etc. How does that work?

The rollup config is here because there is some complexity to make it work and optimize the bundle:
- Monaco contains a lot of code from vscode, and we try bundle vscode code only if this code is not already shipped in monaco
- We then try to treeshake out everything that is not necessary, and rollup needs some help to do so

What surprises you in the rollup config? what else?

**Reply by dlvhdr** [reply]
OK cool, so let me try and upgrade to the latest 1.67 and I'll come back with questions if I have any :)

**Reply by dlvhdr** [reply]
@CGNonofr do you have an example of what you provide `getConfigurationServiceOverride` with?
The README says to provide a `reader` and an `onChange` callbacks.
I'm getting:

```
api.js:14904 Uncaught TypeError: this.configurationService.getConfigurationData is not a function
    at MainThreadConfiguration2._getConfigurationData (api.js:14904:67)
    at new MainThreadConfiguration2 (api.js:14898:45)
    at createExtHostServices (api.js:15140:58)
    at getExtHostServices (api.js:15167:27)
    at Object.registerTextDocumentContentProvider (api.js:15562:53)
    at main.ts:150:6
```

Also, do I need to build `monaco-languageclient` myself and set its `monaco-vscode-api` version to the right one? Because I see it's using 1.68.

**Reply by CGNonofr** [reply]
In the meantime, you can probably play with npm-force-resolutions

Regarding `getConfigurationServiceOverride`, it seems that the configurationService wasn't overriden, do you provide the result of the function to `StandaloneServices.initialize`?
Note that `StandaloneServices.initialize` can only be called once, and creating an editor calls it too (the the third `overrides` parameter). the next calls are ignored.
Also, if a service is used, it cannot be overriden anymore, so `StandaloneServices.initialize` should be called before anything else

I'll add a troubleshoot section in the README

**Reply by dlvhdr** [reply]
> Also, if a service is used, it cannot be overriden anymore, so StandaloneServices.initialize should be called before anything else

Yes my code looks like this:

```ts
StandaloneServices.initialize({
  ...modelEditorService,
  ...getMessageServiceOverride(document.body),
  ...getConfigurationServiceOverride(
    () => {
      console.log("[DEBUG] getConfigurationServiceOverride");
      return "";
    },
    () => {
      return new vscode.Disposable(() => {});
    }
  ),
});

const veloProvider = {
  provideTextDocumentContent: async (uri: monaco.Uri) => {
    const resource = await fetchResource(uri).catch((e) => {
      // StandaloneServices.get(INotificationService).error(e);
      throw e;
    });
    return resource;
  },
};
vscode.workspace.registerTextDocumentContentProvider("velo", veloProvider);
```

I've included the last line because as you can see the stack trace begins with `Object.registerTextDocumentContentProvider`.

I'm gonna try npm resolutions in the meantime.

**Reply by dlvhdr** [reply]
OK using `npm-force-resolutions` solved the problem.
What I did for future reference:

```json
  "resolutions": {
    "@codingame/monaco-vscode-api": "1.67.16"
  },
```

**Reply by CGNonofr** [reply]
Btw, I've just released a breaking change version that simplifies the user configuration api and fix the `getConfigurationData` error when the configurationService is not overriden

**Reply by dlvhdr** [reply]
> Btw, I've just released a breaking change version that simplifies the user configuration api and fix the `getConfigurationData` error when the configurationService is not overriden

Yeah I tried upgrading to 1.67.19 but I'm getting errors:

```
errors.js:12 Uncaught Error: this.configuration.compareAndUpdateDefaultConfiguration is not a function

TypeError: this.configuration.compareAndUpdateDefaultConfiguration is not a function
    at ConfigurationService.onDidDefaultConfigurationChange (configuration.js:1033:43)
    at configuration.js:998:113
    at Listener.invoke (event.js:407:23)
    at Emitter.fire (event.js:551:30)
    at ConfigurationRegistry.registerDefaultConfigurations (configurationRegistry.js:83:40)
    at main.ts:129:6
    at errors.js:12:27
(anonymous) @ errors.js:12
setTimeout (async)
ErrorHandler.unexpectedErrorHandler @ errors.js:10
onUnexpectedError @ errors.js:24
onUnexpectedError @ errors.js:36
fire @ event.js:554
registerDefaultConfigurations @ configurationRegistry.js:83
(anonymous) @ main.ts:129
15:58:51.869 errors.js:12 Uncaught Error: this.updated.fill is not a function

TypeError: this.updated.fill is not a function
    at new FileChangesEvent (services-9299c4ca.js:545:34)
    at configuration.js:1155:98
    at Listener.invoke (event.js:407:23)
    at Emitter.fire (event.js:551:30)
    at configuration.js:2125:26
    at Listener.invoke (event.js:407:23)
    at Emitter.fire (event.js:551:30)
    at updateUserConfiguration (configuration.js:2169:36)
    at main.ts:136:16
    at errors.js:12:27
```

**Reply by CGNonofr** [reply]
I reproduce, there is some other missing polyfills, fix coming

**Reply by CGNonofr** [reply]
You can give a try to the 1.67.20

**Reply by dlvhdr** [reply]
This works, thanks!

**Comment by dlvhdr** [comment]
@CGNonofr 

Regaring the `extHost`, I think I'm just missing the last piece in the puzzle.

> It gathers code from vscode, the extHost classes are the "thing" vscode use between its api and the services

I understand you created this for "impersonating" a vscode extension.
You just hook up things similarly to how vscode does it.
I've looked at the minified code and the last call in the stack before it reaches `monaco.languages` is this inside `extHostLanguageFeatures.js`:

```js
    registerHoverProvider(extension, selector, provider, extensionId) {
        const handle = this._addNewAdapter(new HoverAdapter(this._documents, provider), extension);
        this._proxy.$registerHoverProvider(handle, this._transformDocumentSelector(selector));
        return this._createDisposable(handle);
    }
```

How does this reach monaco in the end?

**Reply by CGNonofr** [reply]
It doesn't use the official monaco api anymore, instead it imports and use the internal registries.

That's why it's now super important to use the exact same version of vscode version that was used to build monaco-editor.

(fortunately, the package.json of monaco-editor includes the commit hash of vscode that was used for the build)
